File: study-project/parser-js/modules/ast/ast.mjs
// @ts-nocheck

/**
 * The expression should have the following EBNF form:
 * ---------------------------------------------------------
 * <expression> ::= <term> { <or> <term> }
 * <term> ::= <factor> { <and> <factor> }
 * <factor> ::= <var> | <not> <factor> | (<expression>)
 * <or>  ::= '|'
 * <and> ::= '&'
 * <not> ::= '!'
 * <var> ::= '[a-zA-Z0-9]*'
 * ---------------------------------------------------------
 */

/**
 * Ast node.
 */
function AstNode() {
  this.evaluate = function evaluate(vars) {};
}

/**
 * Or is the logical OR Operator in an AST.
 *
 * @param lhs
 * @param rhs
 */
function Or(lhs, rhs) {
  this.lhs = lhs;
  this.rhs = rhs;

  this.lhs.evaluateOr() = evaluateOr(vars, this);
  this.rhs.evaluateOr() = evaluateOr(vars, this);
}

/**
 * @param {Map<string, boolean>} vars
 * @param {Or} or
 * @returns Boolean
 */
function evaluateOr(vars, or) {
  return or.lhs.evaluateOr(vars) || or.rhs.evaluateOr(vars);
}

/**
 * And is the logical AND Operator in an AST.
 *
 * @param lhs
 * @param rhs
 */
function And(lhs, rhs) {
  this.lhs = lhs;
  this.rhs = rhs;

  this.lhs.evaluateAnd() = evaluateAnd(vars, this);
  this.rhs.evaluateAnd() = evaluateAnd(vars, this);
}

/**
 * @param {Map<string, boolean>} vars
 * @param {And} and
 * @returns Boolean
 */
function evaluateAnd(vars, and) {
  return and.lhs.evaluateAnd(vars) && and.rhs.evaluateAnd(vars);
}

/**
 * Not is the NOT operator in the AST
 */
function Not() {
  this.lhs.evaluateNot() = evaluateAnd(vars, this);
}

/**
 * @param {Map<string, boolean>} vars
 * @param {Not} not
 * @returns Boolean
 */
function evaluateNot(vars, not) {
  return !not.evaluateNot(vars, not);
}

/**
 * Value is a boolean variable in an AST.
 *
 * @param {string} name
 */
function Value(name) {
  this.name = name;
}

/**
 * @param {Map<string, boolean>} vars
 * @param {Value} value
 * @returns Boolean
 */
function evaluateValue(vars, value) {
  return vars[value.name];
}

File: go-parser/ast/ast.go
package ast

import "fmt"

// Node is the interface to eval an abstract syntax tree (AST).
type Node interface {

	// Eval evaluates the AST. The variables of the expression are set to true or false in the vars map.
	// Missing vars are evaluated to false.
	Eval(vars map[string]bool) bool
}

// Or is the logical OR Operator in an AST.
type Or struct {
	LHS Node
	RHS Node
}

// Eval implements the Node interface.
func (o Or) Eval(vars map[string]bool) bool {
	return o.LHS.Eval(vars) || o.RHS.Eval(vars)
}

func (o Or) String() string {
	return fmt.Sprintf("|(%v,%v)", o.LHS, o.RHS)
}

// And is the logical AND Operator in an AST.
type And struct {
	LHS Node
	RHS Node
}

// Eval implements the Node interface.
func (a And) Eval(vars map[string]bool) bool {
	return a.LHS.Eval(vars) && a.RHS.Eval(vars)
}

func (a And) String() string {
	return fmt.Sprintf("&(%v,%v)", a.LHS, a.RHS)
}

// Not is the NOT operator in the AST.
type Not struct {
	Ex Node
}

// Eval implements the Node interface.
func (n Not) Eval(vars map[string]bool) bool {
	return !n.Ex.Eval(vars)
}

func (n Not) String() string {
	return fmt.Sprintf("!(%v)", n.Ex)
}

// Val is a boolean variable in an AST.
type Val struct {
	Name string
}

// Eval implements the Node interface.
func (v Val) Eval(vars map[string]bool) bool {
	return vars[v.Name] // Missing vars will be evaluated to false.
}

func (v Val) String() string {
	return fmt.Sprintf("'%v'", v.Name)
}
